lib/normalize_grammar.re

```diff
@@ -1,5 +1,62 @@
-open Common;
+module A = Ast_grammar;
+module B = Ast_grammar_normalized;
 
-let normalize = _ast =>
-  raise (Todo);
-  
\ No newline at end of file
+let gensym = () => {
+  /* TODO make it static variable */
+  let static_ident_counter = 1;
+  "intermediate" ++ string_of_int(static_ident_counter)
+}
+
+let rec normalize_to_atom = body => {
+  switch (body) {
+  | A.TOKEN => (B.TOKEN, [])
+  | A.SYMBOL(name)=> (B.SYMBOL(name), [])
+  /* Create intermediate symbol */
+  | _ => {
+    let fresh_ident = gensym();
+    (B.SYMBOL(fresh_ident), [(fresh_ident, fst(normalize_body(body)))]);
+    }
+  }
+}
+and normalize_body = (rule_body) => {
+  switch(rule_body) {
+  | A.TOKEN =>
+    (
+      B.SIMPLE(B.ATOM(B.TOKEN)),
+      []
+    );
+  | A.SYMBOL(name) =>
+    (
+      B.SIMPLE(B.ATOM(B.SYMBOL(name))),
+      []
+    );
+  | A.SEQ(bodies) => {
+      let xs = List.map(normalize_to_atom, bodies);
+      let atoms = List.map(fst, xs);
+      let intermediates = List.flatten(List.map(snd, xs));
+      (B.SIMPLE(B.SEQ(atoms)), intermediates)
+    };
+  | A.CHOICE(bodies) => {
+      let xs = List.map(normalize_to_atom, bodies);
```

Comment:
Here you should call normalize_to_simple, because B.CHOICE take simple, not atoms.
You miss the opportunity to get some CHOICE(SEQ), you'll always get CHOICE(B.ATOM(, which would lead to
less elegant ASTs.